administrator/components/com_templates/Model/TemplateModel.php

```diff
@@ -67,6 +67,159 @@ protected function getFile($path, $name)
+
+			return $temp;
+		}
+		else
+		{
+			return null;
+		}
+	}
+
+	/**
+	 * Method to get all template list.
+	 *
+	 * @return  object  stdClass object
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getTemplateList()
+	{
+		// Get a db connection.
+		$db = Factory::getDbo();
+
+		// Create a new query object.
+		$query = $db->getQuery(true);
+
+		// Select the required fields from the table
+		$query->select(
+			$this->getState(
+				'list.select',
+				'a.extension_id, a.name, a.element, a.client_id'
+			)
+		);
+
+		$query->from($db->quoteName('#__extensions', 'a'))
+			->where($db->quoteName('a.enabled') . ' = 1')
+			->where($db->quoteName('a.type') . ' = ' . $db->quote('template'));
+
+		// Reset the query.
+		$db->setQuery($query);
+
+		// Load the results as a list of stdClass objects.
+		$results = $db->loadObjectList();
+
+		return $results;
+	}
+
+	/**
+	 * Method to get a list of all the core files of override files.
+	 *
+	 * @return  array  A array of all core files.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getCoreList()
+	{
+		// Get list of all templates
+		$templates = $this->getTemplateList();
+
+		// Intialize the array variable to store core file list.
+		$this->coreFileList = array();
+
+		$app    = \JFactory::getApplication();
+
+		foreach ($templates as $template)
+		{
+			$client  = ApplicationHelper::getClientInfo($template->client_id);
+			$element = \JPath::clean($client->path . '/templates/' . $template->element . '/');
+			$path    = \JPath::clean($element . 'html/');
+
+			if (is_dir($path))
+			{
+				$this->prepareCoreFiles($path, $template->client_id, $element, $template->element);
+			}
+			else
+			{
+				$app->enqueueMessage(\JText::_('COM_TEMPLATES_ERROR_TEMPLATE_FOLDER_NOT_FOUND'), 'error');
+
+				return false;
+			}
+		}
+
+		return $this->coreFileList;
+	}
+
+	/**
+	 * Prepare core files.
+	 *
+	 * @param   string   $dir       The path of the directory to scan.
+	 * @param   integer  $client    The client id.
+	 * @param   string   $element   The path of the template element.
+	 * @param   string   $template  The name of template.
+	 *
+	 * @return  array
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function prepareCoreFiles($dir, $client, $element, $template)
+	{
+		$dirFiles = scandir($dir);
+
+		foreach ($dirFiles as $key => $value)
+		{
+			if (!in_array($value, array('.', '..', 'node_modules')))
+			{
+				if (is_dir($dir . $value))
+				{
+					$relativePath = str_replace($element, '', $dir . $value);
+					$this->prepareCoreFiles($dir . $value . '/', $client, $element, $template);
+				}
+				else
+				{
+					$ext           = pathinfo($dir . $value, PATHINFO_EXTENSION);
+					$allowedFormat = $this->checkFormat($ext);
+
+					if ($allowedFormat == true)
+					{
+						$relativePath = str_replace($element, '', $dir);
+						$info = $this->storeFileInfo('/' . $relativePath, $value, $client, $template);
+
+						if($info && $info !== null)
```

Comment:
I think @laoneo means, that there are many if statements and for loops together.

Perhaps you can introduce a new method rather than to use more nested blocks. For example you can put the code in the statement `if (!in_array($value, array('.', '..', 'node_modules')))` into a separate method.

Reply:
@laoneo @astridx I think it was fine because this is a recursion function. Which travel through all files and path of given dir. What do you think ?

Reply:
I have just tried to minimize the nesting. Please look if it is still correct.

```
	/**
	 * Prepare core files.
	 *
	 * @param   string   $dir       The path of the directory to scan.
	 * @param   integer  $client    The client id.
	 * @param   string   $element   The path of the template element.
	 * @param   string   $template  The name of template.
	 *
	 * @return  array
	 *
	 * @since   __DEPLOY_VERSION__
	 */
	public function prepareCoreFiles($dir, $client, $element, $template)
	{
		$dirFiles = scandir($dir);

		// For all files in templates html folder
		foreach ($dirFiles as $value)
		{
			// Exclude directories node_module, . and ..
			if (in_array($value, array('.', '..', 'node_modules')))
			{
				continue;
			}

			$ext = pathinfo($dir . $value, PATHINFO_EXTENSION);
			$allowedFormat = $this->checkFormat($ext);

			// Continue if file and file extension is not allowed.
			if (!is_dir($dir . $value) && $allowedFormat !== true)
			{
				continue;
			}

			$this->info = new \stdClass;

			// If directory call recursion otherwise create info.
			if (is_dir($dir . $value))
			{
				$relativePath = str_replace($element, '', $dir . $value);
				$this->prepareCoreFiles($dir . $value . '/', $client, $element, $template);
			}
			else
			{
				$relativePath = str_replace($element, '', $dir);
				$this->info = $this->storeFileInfo('/' . $relativePath, $value, $client, $template);
			}

			// If info save it in the coreFileList.
			if ($this->info && $this->info !== null)
			{
				$this->coreFileList[] = $this->info;
			}
		}

		return true;
	}

```

Reply:
Much easier to read like that for an outsider.